apps/BlockedStencil/BlockedStencil.cpp

```diff
@@ -0,0 +1,151 @@
+/**
+ * A simple stencil computation that computes the sum of each point 
+ * and its four direct neighbours in a 2D grid.
+ * The implementation accesses memory in an aligned way and uses shared memory.
+ * Each thread computes a single element in the output buffer.
+ *
+ * Author: Paul Metzger 
+ */
+
+#include <NoCL.h>
+
+#define DEBUG false
+
+void populate_in_buf(int *in_buf, int x_size, int y_size) {
+  for (int y = 0; y < y_size; ++y) {
+    for (int x = 0; x < x_size; ++x) in_buf[y * x_size + x] = x * y;
+  }
+}
+
+// Generate a 'golden output' to check if the output computed
+// by the GPU kernel is correct.
+void generate_golden_output(int *in_buf, int *golden_out, int x_size, int y_size) {
+  for (int y = 0; y < y_size; ++y) {
+    for (int x = 0; x < x_size; ++x) {
+      const int ind = y * x_size + x;
+
+      int result = in_buf[ind];
+      if (x < x_size - 1) result += in_buf[y * x_size + x + 1];
+      if (x > 0)          result += in_buf[y * x_size + x - 1];
+      if (y < y_size - 1) result += in_buf[(y + 1) * x_size + x];
+      if (y > 0)          result += in_buf[(y - 1) * x_size + x];
+      golden_out[ind] = result;
+    }
+  }
+}
+
+// Check if the results computed by the GPU kernel match
+// the golden output.
+bool check_output(int *out_buf, int *golden_buf, int buf_size) {
+  for (int i = 0; i < buf_size; ++i) {
+    if (out_buf[i] != golden_buf[i]) {
+      puts("Detected an error at index: "); puthex(i); putchar('\n');
+      puts("Expected value: "); puthex(golden_buf[i]); putchar('\n');
+      puts("Computed value: "); puthex(out_buf[i]);    putchar('\n');
+      return false;
+    }
+  }
+  return true;
+}
+
+struct SimpleStencil : Kernel {
+  int x_size = 0;
+  int y_size = 0;
+  int *out_buf, *in_buf;
+
+  void kernel() {
+    const int x = blockIdx.x * blockDim.x + threadIdx.x;
+    const int y = blockIdx.y * blockDim.y + threadIdx.y;
+    const int ind = y * x_size + x;
+
+    // Load values into local memory
+    auto c = shared.array<int, SIMTWarps, SIMTLanes>();
```

Comment:
Could do:

```
auto c = shared.array<int>(blockDim.y, blockDim.x);
```

That would be more in line with the other samples, but it's slightly slower.  Up to you.

Reply:
I've changed it to what you suggested.